stan/math/rev/fun/add.hpp

```diff
@@ -0,0 +1,538 @@
+ *
+ * @param[in] A matrix
+ * @param[in] B matrix
+ * @return Sum of two matrices
+ */
+template <typename Mat1, typename Mat2,
+          require_all_eigen_t<Mat1, Mat2>* = nullptr,
+          require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
+auto add(const Mat1& A, const Mat2& B) {
+  using Ta = value_type_t<Mat1>;
+  using Tb = value_type_t<Mat2>;
+  constexpr int R = Mat1::RowsAtCompileTime;
+  constexpr int C = Mat1::ColsAtCompileTime;
+  check_matching_dims("add", "A", A, "B", B);
+  check_not_nan("add", "A", A);
+  check_not_nan("add", "B", B);
+
+  // Manage memory in arena allocator
+  add_mat_vari<Ta, Tb, R, C>* baseVari = new add_mat_vari<Ta, Tb, R, C>(A, B);
+  Eigen::Matrix<var, R, C> AplusB_v(A.rows(), A.cols());
+  AplusB_v.vi()
+      = Eigen::Map<matrix_vi>(&baseVari->variRefAplusB_[0], A.rows(), A.cols());
+
+  return AplusB_v;
+}
+
+/**
+ * This is a subclass of the vari class for matrix
+ * addition A + c, where A is a N by M matrix and
+ * c is a scalar.
+ *
+ * The class stores the structure of each matrix,
+ * the double values of A and c, and pointers to
+ * the varis for A and c if A or c is a var. It
+ * also instatiates and stores pointers to varis
+ * for all elements of A + c.
+ *
+ * @tparam Ta type of elements in matrix A
+ * @tparam Tc type of elements in matrix B
+ * @tparam R number of rows (common to A and B)
+ * @tparam C number of columns (common to A and B)
+ */
+
+template <typename Ta, typename Tc, int R, int C>
+class add_mat_scal_vari : public vari {
+ public:
+  int rows_;
+  int cols_;
+  int size_;
+  double* Ad_;
+  double cd_;
+  vari** variRefA_;
+  vari* variRefc_;
+  vari** variRefAplusc_;
+
+  /**
+   * Constructor for add_mat_vari.
+   *
+   * All memory is allocated in ChainableStack's
+   * stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled by the second argument to
+   * vari's constructor.
+   *
+   * @param A matrix
+   * @param c scalar
+   **/
+  add_mat_scal_vari(const Eigen::Matrix<Ta, R, C>& A, const Tc c)
+      : vari(0.0),
+        rows_(A.rows()),
+        cols_(A.cols()),
+        size_(A.size()),
+        Ad_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        cd_(c.val()),
+        variRefA_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)),
+        variRefc_(c.vi_),
+        variRefAplusc_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)) {
+    using Eigen::Map;
+    Map<matrix_vi>(variRefA_, rows_, cols_) = A.vi();
+    Map<matrix_d> Ad(Ad_, rows_, cols_);
+    double cd = c.val();
+    Ad = A.val();
+    Map<matrix_vi>(variRefAplusc_, rows_, cols_)
+        = (Ad.array() + cd).matrix().eval().unaryExpr([](double x) {
+            return new vari(x, false);
+          });
+  }
```

Comment:
```suggestion
    Map<matrix_d> Ad(Ad_, rows_, cols_);
    Ad = A.val();
    Map<matrix_vi>(variRefAplusc_, rows_, cols_).array()
        = (Ad.array() + cd_).unaryExpr([](double x) {
            return new vari(x, false);
          });
  }
```

This cuts out the unnecessary ```cd``` temporary, and the creation of the temporary matrix caused by ```.matrix().eval()```